src/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs

```diff
@@ -18,57 +18,82 @@ internal partial class DiagnosticAnalyzerService
+        public async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsForIdsAsync(Project project, DocumentId? documentId, ImmutableHashSet<string>? diagnosticIds, Func<DiagnosticAnalyzer, bool>? shouldIncludeAnalyzer, bool includeLocalDocumentDiagnostics, bool includeNonLocalDocumentDiagnostics, CancellationToken cancellationToken)
+        {
+            return await ProduceProjectDiagnosticsAsync(
+                project, diagnosticIds, shouldIncludeAnalyzer,
+                // Ensure we compute and return diagnostics for both the normal docs and the additional docs in this
+                // project if no specific document id was requested.
+                documentId != null ? [documentId] : [.. project.DocumentIds, .. project.AdditionalDocumentIds],
+                includeLocalDocumentDiagnostics,
+                includeNonLocalDocumentDiagnostics,
+                // return diagnostics specific to one project or document
+                includeProjectNonLocalResult: documentId == null,
+                cancellationToken).ConfigureAwait(false);
+        }
 
-        public Task<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsForIdsAsync(Project project, ImmutableHashSet<string>? diagnosticIds, Func<DiagnosticAnalyzer, bool>? shouldIncludeAnalyzer, bool includeNonLocalDocumentDiagnostics, CancellationToken cancellationToken)
-            => new DiagnosticGetter(this, project, documentId: null, diagnosticIds, shouldIncludeAnalyzer, includeLocalDocumentDiagnostics: false, includeNonLocalDocumentDiagnostics).GetProjectDiagnosticsAsync(cancellationToken);
+        public async Task<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsForIdsAsync(
+            Project project,
+            ImmutableHashSet<string>? diagnosticIds,
+            Func<DiagnosticAnalyzer, bool>? shouldIncludeAnalyzer,
+            bool includeNonLocalDocumentDiagnostics,
+            CancellationToken cancellationToken)
+        {
+            return await ProduceProjectDiagnosticsAsync(
+               project, diagnosticIds, shouldIncludeAnalyzer,
+               documentIds: [],
+               includeLocalDocumentDiagnostics: false,
+               includeNonLocalDocumentDiagnostics: includeNonLocalDocumentDiagnostics,
+               includeProjectNonLocalResult: true,
+               cancellationToken).ConfigureAwait(false);
+        }
 
-        private sealed class DiagnosticGetter(
-            DiagnosticIncrementalAnalyzer owner,
+        private async Task<ImmutableArray<DiagnosticData>> ProduceProjectDiagnosticsAsync(
             Project project,
-            DocumentId? documentId,
             ImmutableHashSet<string>? diagnosticIds,
             Func<DiagnosticAnalyzer, bool>? shouldIncludeAnalyzer,
+            IReadOnlyList<DocumentId> documentIds,
             bool includeLocalDocumentDiagnostics,
-            bool includeNonLocalDocumentDiagnostics)
```

Comment:
general idea is that instead of making tehse nested types, which have to grab all this data, just to store fields, just to use in methods within the type, we just use local functions.